src/shape.cpp

```diff
@@ -48,20 +48,20 @@ struct shape_impl
 
     shape_impl(shape::type_t t) : m_type(t), m_lens({1}), m_strides({0}), m_standard(true)
     {
-        assert(t != shape::tuple_type);
+        assert(t != shape::tuple_type && t != shape::uint4_type && t != shape::int4_type);
     }
 
     shape_impl(shape::type_t t, std::vector<std::size_t> l)
         : m_type(t), m_lens(std::move(l)), m_standard(true)
     {
-        assert(t != shape::tuple_type);
+        assert(t != shape::tuple_type && t != shape::uint4_type && t != shape::int4_type);
```

Comment:
This assert shouldn't be changed here. It should be valid to construct a shape of int4_type otherwise whats the point of adding this operator.

Reply:
Thanks. This one is definitely required.

Reply:
What do you mean?

Reply:
I meant you are correct. This constructor code, as you suggest, is definitely required for correct functioning. 
(I had just quickly hacked those places to make sure the CI would pass the compile stage -- as some switch statements etc. were failing.) Thanks.